RFE-2782: Create edge compute pool to support AWS Local Zones

mtulio opened the pull request:
The goal of this PR is to add support to [AWS Local Zones](https://aws.amazon.com/about-aws/global-infrastructure/localzones/) when installing a cluster in an existing VPC. It allows the creation of a compute pool named `edge`, alongside the `worker`, on `install-config.yaml`. The installer will generate MachineSet manifests with `NoSchedule` taint to deploy the instance on edge locations.

This implementation is Phase 1 of the [Enhancement Proposal #1232](https://github.com/openshift/enhancements/pull/1232), which describes in detail the change request.

Nodes are considered `edge` when the availability zone  [`zoneType`](https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_AvailabilityZone.html) is `local-zone`. The `edge` compute pool is valid only for the AWS Platform.

It is not the goal of this PR to create the Subnets on Local Zones locations. This is part of Phase 2.

Note: This PR will change the MTU of ClusterNetwork when edge pool for AWS is defined. FYI https://github.com/openshift/openshift-docs/pull/55327

Refereces:
- https://issues.redhat.com/browse/RFE-2782
- https://issues.redhat.com/browse/SPLAT-636
- https://github.com/openshift/enhancements/pull/1232


List of teams that would be interested in reviewing this PR:

- [x] SPLAT (@rvanderp3 )
- [x] Installer (@patrickdillon  @sadasu )
- [ ] Network Team: Cluster network MTU Changes (@abhat @jcaamano @trozet )
- [ ] Cloud: MachineSet changes (@JoelSpeed @elmiko )

mtulio commented:
/hold

mtulio commented:
Tests with the current PR on an IPI installation in existing VPC in `us-west-2` with one subnet for each Local Zones available in that region.

Nodes:
```
$ oc get nodes
NAME                                         STATUS   ROLES                  AGE   VERSION
ip-10-0-13-183.us-west-2.compute.internal    Ready    worker                 28m   v1.24.0+8c7c967
ip-10-0-13-35.us-west-2.compute.internal     Ready    control-plane,master   36m   v1.24.0+8c7c967
ip-10-0-16-27.us-west-2.compute.internal     Ready    control-plane,master   36m   v1.24.0+8c7c967
ip-10-0-18-192.us-west-2.compute.internal    Ready    worker                 28m   v1.24.0+8c7c967
ip-10-0-194-163.us-west-2.compute.internal   Ready    edge,worker            18m   v1.24.0+8c7c967
ip-10-0-196-196.us-west-2.compute.internal   Ready    edge,worker            26m   v1.24.0+8c7c967
ip-10-0-202-81.us-west-2.compute.internal    Ready    edge,worker            18m   v1.24.0+8c7c967
ip-10-0-205-248.us-west-2.compute.internal   Ready    edge,worker            18m   v1.24.0+8c7c967
ip-10-0-21-119.us-west-2.compute.internal    Ready    worker                 28m   v1.24.0+8c7c967
ip-10-0-211-241.us-west-2.compute.internal   Ready    edge,worker            25m   v1.24.0+8c7c967
ip-10-0-213-135.us-west-2.compute.internal   Ready    edge,worker            18m   v1.24.0+8c7c967
ip-10-0-218-87.us-west-2.compute.internal    Ready    edge,worker            21m   v1.24.0+8c7c967
ip-10-0-22-68.us-west-2.compute.internal     Ready    control-plane,master   35m   v1.24.0+8c7c967

```

Machines:
> Note: for Los Angeles (`us-west-2-lax-1{a,b}`) location the preferred instance was `m5.xlarge` as it has more precedence than others, but is not offered across other locations. See more on the [EP discussion](https://github.com/openshift/enhancements/pull/1232/files#r979980109) and [AWS Local Zones Features Documentation](https://github.com/openshift/enhancements/pull/1232/files#r979980109).

```
$ oc get machines -n openshift-machine-api
NAME                                                PHASE     TYPE          REGION      ZONE               AGE
ipi-zzz-test124-jwcpd-edge-us-west-2-den-1a-vrldx   Running   c5d.2xlarge   us-west-2   us-west-2-den-1a   32m
ipi-zzz-test124-jwcpd-edge-us-west-2-las-1a-4tgnc   Running   c5d.2xlarge   us-west-2   us-west-2-las-1a   32m
ipi-zzz-test124-jwcpd-edge-us-west-2-lax-1a-jfkc7   Running   m5.xlarge     us-west-2   us-west-2-lax-1a   32m
ipi-zzz-test124-jwcpd-edge-us-west-2-lax-1b-4tdk8   Running   m5.xlarge     us-west-2   us-west-2-lax-1b   32m
ipi-zzz-test124-jwcpd-edge-us-west-2-pdx-1a-9vmgk   Running   c5d.2xlarge   us-west-2   us-west-2-pdx-1a   32m
ipi-zzz-test124-jwcpd-edge-us-west-2-phx-1a-6hnk2   Running   c5d.2xlarge   us-west-2   us-west-2-phx-1a   32m
ipi-zzz-test124-jwcpd-edge-us-west-2-sea-1a-zjtwb   Running   c5d.2xlarge   us-west-2   us-west-2-sea-1a   32m
ipi-zzz-test124-jwcpd-master-0                      Running   m6i.xlarge    us-west-2   us-west-2b         36m
ipi-zzz-test124-jwcpd-master-1                      Running   m6i.xlarge    us-west-2   us-west-2c         36m
ipi-zzz-test124-jwcpd-master-2                      Running   m6i.xlarge    us-west-2   us-west-2a         36m
ipi-zzz-test124-jwcpd-worker-us-west-2a-vkdbl       Running   m6i.xlarge    us-west-2   us-west-2a         32m
ipi-zzz-test124-jwcpd-worker-us-west-2b-jrfss       Running   m6i.xlarge    us-west-2   us-west-2b         32m
ipi-zzz-test124-jwcpd-worker-us-west-2c-4jlnt       Running   m6i.xlarge    us-west-2   us-west-2c         32m

```

The manifests of the following hypothetical configurations were also rendered:

```
TEST_ID-NAME				 RESULT	 INST_MANIFEST_RC
--------				 --	 -- 	 --
t00_01-aws-exist_vpc_workers-empty	 OK	 OK[0]		
t00_01-aws-exist_vpc_workers-pool_edge	 OK	 OK_ERR[3]	
t01-aws-exist_vpc_all-pool_empty	 OK	 OK[0]		
t02-aws-exist_vpc_all-pool_worker	 OK	 OK[0]		
t03-aws-exist_vpc_all-pool_edge_only	 OK	 OK[0]		
t04-aws-exist_vpc_all-pool_wk_edge	 OK	 OK[0]		
t05-aws-exist_vpc_all-pool_edge_inst	 OK	 OK[0]		
t06-aws-exist_vpc_all-pool_edge_ebs	 OK	 OK[0]		
t07-aws-exist_vpc_all-pool_edge_zones	 OK	 OK[0]		
t08-aws-exist_vpc_all-pool_edge_repl	 OK	 OK[0]		
t09-aws-exist_vpc_edge_net-pool_edge	 OK	 OK_ERR[3]	
t20_01-plat_none-default__config	 OK	 OK[0]		
t20_02-plat_none-pool_worker_edge	 OK	 OK_ERR[3]	
t30_01-plat_azure-default_config	 OK	 OK[0]		
t30_02-plat_azure-pool_worker_only	 OK	 OK[0]		
t30_03-plat_azure-pool_worker_edge	 OK	 OK_ERR[3]	
t30_04-plat_azure-pool_edge_only	 OK	 OK_ERR[3]	
```

barbacbd commented on pkg/asset/installconfig/aws/subnet.go:
```diff
@@ -99,38 +145,62 @@ func subnets(ctx context.Context, session *session.Session, region string, ids [
+		meta.ZoneType = *availabilityZones[meta.Zone].ZoneType
+		meta.ZoneGroup = *availabilityZones[meta.Zone].GroupName
+
+		if meta.ZoneType == awstypes.LocalZoneType {
```
nit: Instead of `continue` in several places, could this be altered to use if/if-else/else ?

mtulio replied:
@barbacbd that's a good idea. Changed.

barbacbd commented on pkg/asset/installconfig/aws/validation.go:
```diff
@@ -49,12 +49,32 @@ func Validate(ctx context.Context, meta *Metadata, config *types.InstallConfig)
+					// TODO create a test for this error
+					return errors.New(field.Required(fldPath, "invalid install config. There is no valid subnets for edge machine pool.").Error())
+				}
+				break
```
is the `break` required?

barbacbd commented on pkg/asset/installconfig/aws/validation.go:
```diff
@@ -49,12 +49,32 @@ func Validate(ctx context.Context, meta *Metadata, config *types.InstallConfig)
+			switch compute.Name {
+			// Pool's specific validation
+			// Edge Compute Pool: AWS Local Zones is valid only when installing in existing VPC.
+			case types.MachinePoolEdgeRoleName:
```
Is there an intention to add more cases? If not then this could be an if right ?

mtulio replied:
No, if we haven't any checks specific for `worker` compute type. As it's not the case, I will replace the `switch` here. Thanks!

barbacbd commented on pkg/asset/installconfig/aws/validation.go:
```diff
@@ -49,12 +49,32 @@ func Validate(ctx context.Context, meta *Metadata, config *types.InstallConfig)
+				}
+				edgeSubnets, err := meta.EdgeSubnets(ctx)
+				if err != nil {
+					allErrs = append(allErrs, field.Invalid(fldPath, edgeSubnets, err.Error()))
```
Is it possible for this error message to be skipped when the len(edgeSubnets) <= 0?

mtulio replied:
Yes, it seems. I wonder if it will be better to add the `return allErrs.ToAggregate()` here. wdyt?

mtulio replied:
Changed [here](https://github.com/openshift/installer/pull/6371/commits/ceedde03d5151c7b68457a2e3ef7304553770141#diff-bab1b8072f369df53e27d65607145fb93caabd3a86fdefbcafb176e823726069R69). Let me know if it's what you were thinking.

mtulio commented:
@patrickdillon would you mind taking a look at this PR?

- `ci/prow/yaml-lint` is not related to the changes here, but the path seems to check changes from the whole path `/upi/*`. If you want, I can create another PR fixing those YAML lint items.
- I am looking for opportunities to increase the tests coverage
- I am looking at the failing in integrated tests, it's not clear if any of it is related to the changes here. I am open to hearing from you.

/retest-required

barbacbd commented on pkg/types/validation/installconfig.go:
```diff
@@ -428,14 +428,33 @@ func validateControlPlane(platform *types.Platform, pool *types.MachinePool, fld
 
+func validateComputeEdge(platform *types.Platform, p *types.MachinePool, fldPath *field.Path, pfld *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	switch platform.Name() {
```
Is this meant to expand or can this really just be an If for now?

mtulio replied:
> for now
for now, yes. Edge pool is supported only in AWS at this moment.

barbacbd commented on pkg/types/defaults/installconfig.go:
```diff
@@ -71,8 +71,15 @@ func SetInstallConfigDefaults(c *types.InstallConfig) {
+	defaultComputePoolUndefined := true
+	for _, compute := range c.Compute {
+		if compute.Name == types.MachinePoolComputeRoleName {
+			defaultComputePoolUndefined = false
```
nit: break early?

patrickdillon commented on upi/aws/cloudformation/01.99_net_local-zone.yaml:
```diff
@@ -0,0 +1,50 @@
+    Description: Subnet IDs of the public subnets.
+    Value:
+      !Join [
+        "",
```
the `yaml-lint` failure looks to be related to this line:

```
  48:11     error    wrong indentation: expected 8 but found 10  (indentation)
```
This looks simple enough you could put it on a single line.

mtulio replied:
You are right, I missed that line. Fixed. Thanks!

patrickdillon commented:
> @patrickdillon would you mind taking a look at this PR?
> 
> * `ci/prow/yaml-lint` is not related to the changes here, but the path seems to check changes from the whole path `/upi/*`. If you want, I can create another PR fixing those YAML lint items.

There are a lot of unrelated warnings, but the error is from a change in this PR:

```
./upi/aws/cloudformation/01.99_net_local-zone.yaml
  18:121    warning  line too long (144 > 120 characters)  (line-length)
  48:11     error    wrong indentation: expected 8 but found 10  (indentation)
```
I left a comment with a suggestion.

mtulio commented:
@patrickdillon @barbacbd thanks for your comments. I hope I have covered all the items on my last commit. Thanks!

mtulio commented:
> @patrickdillon would you mind taking a look at this PR?
> 
> * `ci/prow/yaml-lint` is not related to the changes here, but the path seems to check changes from the whole path `/upi/*`. If you want, I can create another PR fixing those YAML lint items.
> * I am looking for opportunities to increase the tests coverage
> * I am looking at the failing in integrated tests, it's not clear if any of it is related to the changes here. I am open to hearing from you.
> 
> /retest-required

Items reviewed!

jstuever commented:
/cc @patrickdillon 
/uncc

patrickdillon reviewed (commented):
The commit structure makes this PR a little challenging to review. I believe there are changes which are introduced in one commit and then changed in a subsequent commit. Would it be possible to rebase this and separate the commits into logical units?

mtulio commented:
> The commit structure makes this PR a little challenging to review. I believe there are changes which are introduced in one commit and then changed in a subsequent commit. Would it be possible to rebase this and separate the commits into logical units?

@patrickdillon , yes, for sure. Let us close Phase-1 on the Enhancement, so I can work again on this PR. I will convert it to a draft until we finish the first part of the discussion there.

mtulio commented:
> The commit structure makes this PR a little challenging to review. I believe there are changes which are introduced in one commit and then changed in a subsequent commit. Would it be possible to rebase this and separate the commits into logical units?

@patrickdillon PR commits were isolated into different units. Appreciate your feedback.

mtulio commented:
@patrickdillon @makentenza @rvanderp3 @jcpowermac this PR is ready for review. PTAL?
We are heading planning this for 4.13.
 Thanks!
/assign @patrickdillon

mtulio commented:
> @patrickdillon @makentenza @rvanderp3 @jcpowermac this PR is ready for review. PTAL?
> 
> We are heading planning this for 4.13.
> 
>  Thanks!
> 
> /assign @patrickdillon 

Folks, i am still investigating some CI failures exclusively from this PR that passed from my eyes since my last review. This PR will be ready for review on next Monday.

rvanderp3 commented on pkg/asset/manifests/network.go:
```diff
@@ -7,18 +7,28 @@ import (
 
+const (
+	// Cluster Network MTU for AWS Local Zone deployments on edge machine pools
+	ovnKNetworkMtuEdge   = 1200
+	ocpSDNNetworkMtuEdge = 1250
```
perhaps you could type the constant to avoid having to convert when the variable is used.  
`ovnKNetworkMtuEdge  uint32 = 1200
ocpSDNNetworkMtuEdge uint32 = 1250`

rvanderp3 commented on pkg/asset/installconfig/aws/availabilityzones.go:
```diff
@@ -6,11 +6,12 @@ import (
 	"github.com/aws/aws-sdk-go/aws"
 	"github.com/aws/aws-sdk-go/aws/session"
 	"github.com/aws/aws-sdk-go/service/ec2"
+	awstypes "github.com/openshift/installer/pkg/types/aws"
```
i think the linter is going to want the installer package in its own grouping

mtulio replied:
Thanks. yeah, I realized many other lint and fmt issues after rebase. I am addressing this today.

rvanderp3 commented on pkg/asset/installconfig/aws/subnet.go:
```diff
@@ -9,12 +9,16 @@ import (
 	"github.com/aws/aws-sdk-go/aws"
 	"github.com/aws/aws-sdk-go/aws/session"
 	"github.com/aws/aws-sdk-go/service/ec2"
+	awstypes "github.com/openshift/installer/pkg/types/aws"
```
i think the linter is going to want the installer package in its own grouping

rvanderp3 reviewed (commented):
a couple of nits, but LGTM

mtulio commented:
/retest-required

mtulio commented:
/test e2e-aws-ovn
/test e2e-azure-ovn
/test e2e-gcp-ovn

mtulio commented:
Adding a list of teams that would be interested in reviewing this PR:
- [ ] SPLAT (@rvanderp3 )
- [ ] Installer @patrickdillon 
- [ ] Network Team: Cluster network MTU Changes (@ ?)
- [ ] Cloud: MachineSet changes (@JoelSpeed @elmiko )

@patrickdillon could you please advise anyone from Network team?

@JoelSpeed @elmiko would you mind to take a look into this PR to certify the changes described on EP https://github.com/openshift/enhancements/pull/1232 to introduce the `edge` machine pool when installing in AWS Local Zones?

mtulio commented:
/test e2e-aws-ovn

mtulio commented:
/test e2e-aws-ovn

mtulio commented on docs/user/aws/install_existing_vpc_local-zones.md:
````diff
@@ -0,0 +1,427 @@
+- Create the Stack
+
+```bash
+INSTALER_URL="file:///home/mtulio/go/src/github.com/mtulio/installer"
+#INSTALER_URL="https://raw.githubusercontent.com/mtulio/installer/master"
````
```suggestion
INSTALER_URL="https://raw.githubusercontent.com/openshift/installer/master"
```

mtulio commented on docs/user/aws/install_existing_vpc_local-zones.md:
````diff
@@ -0,0 +1,427 @@
+- Create the Stack for Local Zone subnet `us-west-2-lax-1a`
+
+```bash
+INSTALER_URL="file:///home/mtulio/go/src/github.com/mtulio/installer"
+#INSTALER_URL="https://raw.githubusercontent.com/mtulio/installer/master"
````
```suggestion
INSTALER_URL="https://raw.githubusercontent.com/openshift/installer/master"
```

mtulio commented on docs/user/aws/install_existing_vpc_local-zones.md:
````diff
@@ -0,0 +1,427 @@
+7
+```
+
+##### Creating the subnet using AWS CLI <a name="create-network-subnet-res-cli"></a>
+
+TBD/Is it needed?
````
Already described above
```suggestion
```

mtulio commented on docs/user/aws/install_existing_vpc_local-zones.md:
````diff
@@ -0,0 +1,427 @@
+        ParameterKey=PublicRouteTableId,ParameterValue=${PUBLIC_RTB_ID}
+```
+
+- Wait for the stack be created `StackStatus=CREATE_COMPLETE`
+
+```bash
+aws cloudformation describe-stacks \
+  --region ${AWS_REGION} \
+  --stack-name ${SUBNET_NAME}
+```
+
+- Wait for the stack execution be finished: `StackStatus=CREATE_COMPLETE`.
````
```suggestion
- Wait for the stack to be created `StackStatus=CREATE_COMPLETE`

```bash
aws cloudformation wait stack-create-complete \
  --region ${AWS_REGION} \
  --stack-name ${SUBNET_NAME}
```

```

mtulio commented on docs/user/aws/install_existing_vpc_local-zones.md:
````diff
@@ -0,0 +1,427 @@
+        ParameterKey=AvailabilityZoneCount,ParameterValue=${VPC_SUBNETS_COUNT}
+```
+
+- Check the Stack status:
+
+```bash
+aws cloudformation describe-stacks \
+    --region ${AWS_REGION} \
+    --stack-name ${CLUSTER_NAME}-vpc 
+```
+
+- Wait for the stack execution be finished: `StackStatus=CREATE_COMPLETE`.
````
```suggestion
- Wait for the stack execution to be created: `StackStatus=CREATE_COMPLETE`

```bash
aws cloudformation wait stack-create-complete \
    --region ${AWS_REGION} \
    --stack-name ${CLUSTER_NAME}-vpc 
```

```

mtulio commented on docs/user/aws/install_existing_vpc_local-zones.md:
```diff
@@ -0,0 +1,427 @@
+
+## Install the cluster <a name="install-cluster"></a>
+
+To install the cluster in existing VPC with subnets in Local Zones, you should:
+- generate the install-config.yaml, you provide yours
+- add the subnet IDs setting the option `platform.aws.subnets`
+- (option) customize the Edge Machine Pool
```
```suggestion
To install the cluster in existing VPC with subnets in Local Zones, you should:
- generate the `install-config.yaml`, or provide yours
- add the subnet IDs by setting the option `platform.aws.subnets`
- (optional) customize the `edge` compute pool
```

mtulio commented on docs/user/aws/install_existing_vpc_local-zones.md:
```diff
@@ -0,0 +1,427 @@
+
+### Setting up the Edge Machine Pool <a name="create-config-edge-pool"></a>
+
+Version 4.12 e earlier introduces a new Machine Pool designed for the edge locations.
+The Edge Machine Pool configuration is common between edge locations, but due to the limitation
+of resources (Instance Types and Sizes) of the Local Zone, the default instance type created
+may vary from the traditional worker pool.
```
```suggestion
Version 4.12 or later introduces a new compute pool named `edge` designed for remote zones.
The `edge` compute pool configuration is common between AWS Local Zone locations, but due to the limitation
of resources (Instance Types and Sizes) of the Local Zone, the default instance type created
may vary from the traditional worker pool.
```

mtulio commented on docs/user/aws/install_existing_vpc_local-zones.md:
```diff
@@ -0,0 +1,427 @@
+- `m5.xlarge`
+- `c5d.2xlarge`
+
+The Edge Machine Pool will also create new labels for the Machines to help developers to
```
```suggestion
The `edge`compute pool will also create new labels to help developers to
```

mtulio commented on docs/user/aws/install_existing_vpc_local-zones.md:
```diff
@@ -0,0 +1,427 @@
+    - `zone_type=local-zone`
+    - `zone_group=<Local Zone Group>`
+
+Finally, the MachineSets created on the Edge Machine Pool have `NoSchedule` taint to avoid the
+normal workloads spread out on those machines, and only user workloads will be allowed to run
+when the tolerations are defined on the pod spec.
+
+By default, the Edge Machine Pools will be created only when AWS Local Zone subnet IDs are added
+to the list of `platform.aws.subnets`.
```
```suggestion
Finally, the Machine Sets created by the `edge` compute pool have `NoSchedule` taint to avoid the
regular workloads spread out on those machines, and only user workloads will be allowed to run
when the tolerations are defined on the pod spec.

By default, the `edge` compute pool will be created only when AWS Local Zone subnet IDs are added
to the list of `platform.aws.subnets`.
```

mtulio reviewed (commented):
Reviewing the user documentation for phase-1, tracked by https://issues.redhat.com/browse/OSDOCS-5240

mtulio commented on pkg/asset/machines/aws/machines.go:
```diff
@@ -153,61 +168,74 @@ func Machines(clusterID string, region string, subnets map[string]string, pool *
 		BlockDevices: []machineapi.BlockDeviceMappingSpec{
 			{
 				EBS: &machineapi.EBSBlockDeviceSpec{
-					VolumeType: pointer.StringPtr(root.Type),
-					VolumeSize: pointer.Int64Ptr(int64(root.Size)),
-					Iops:       pointer.Int64Ptr(int64(root.IOPS)),
-					Encrypted:  pointer.BoolPtr(true),
-					KMSKey:     machineapi.AWSResourceReference{ARN: pointer.StringPtr(root.KMSKeyARN)},
```
> Is this actually changing anything?

@JoelSpeed No, just providing the input structure to avoid the long amount of arguments.

> Why not use the pointer utils?

I would, but the linter job was reporting [as deprecated](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_installer/6371/pull-ci-openshift-installer-master-golint/1623845916333051904#1:build-log.txt%3A96-115):

```
A1019: pointer.Int64Ptr is deprecated: Use Int64 instead. (staticcheck)
					Iops:       pointer.Int64Ptr(int64(in.root.IOPS)),
					            ^
pkg/asset/machines/aws/machines.go:190:55: SA1019: pointer.StringPtr is deprecated: Use String instead. (staticcheck)
					KMSKey:     machineapi.AWSResourceReference{ARN: pointer.StringPtr(in.root.KMSKeyARN)},
(...)
```

Thoughts?

mtulio replied:
FYI https://github.com/kubernetes/utils/pull/249

JoelSpeed replied:
I think they just renamed a bunch, so it's `pointer.String` no instead of `pointer.StringPtr`, they basically dropped `Ptr` from all the different function names

mtulio replied:
@JoelSpeed you're completely right. That change passed to my eye. I just fixed that on [this review](https://github.com/openshift/installer/compare/cff35ba7dfe2b78984f49b73bfc20d11419fbe3c..c4d652e3d83bb4821eff4ed9b0a834c69c22ff1e). Thanks

mtulio commented on pkg/types/aws/defaults/platform.go:
```diff
@@ -22,19 +22,23 @@ var (
+// c5d.2xlarge is in the most locations of availability for Local Zone offerings.
+// https://aws.amazon.com/about-aws/global-infrastructure/localzones/features
+// https://aws.amazon.com/ec2/pricing/on-demand/
+func InstanceTypes(region string, arch types.Architecture) []string {
```
The function [`awsDefaultMachineTypes()`](awsDefaultMachineTypes) had hardcode the size `xlarge`, assuming all nodes should be this one, which is not true [as per our discussion on EP](ep-zones), there are some zones that will not support any of `m6i.xlarge` or `m5.xlarge`, and should choose the instance c5d.**2xlarge**.

[awsDefaultMachineTypes]:https://github.com/openshift/installer/pull/6371/files#diff-40838b353833f830328df988f735eb9f51a2a1c113bd8ce0254c91857e6787eeL185-L189
[ep-local-zones]:https://github.com/openshift/enhancements/pull/1232

JoelSpeed replied:
Broken link on the EP discussion but makes sense

mtulio commented on pkg/asset/machines/aws/machinesets.go:
```diff
@@ -32,28 +34,62 @@ func MachineSets(clusterID string, region string, subnets map[string]string, poo
+			}
+			nodeLabels = map[string]string{
+				"node-role.kubernetes.io/edge": "",
+				"zone_type":                    subnet.ZoneType,
+				"zone_group":                   subnet.ZoneGroup,
```
That's a good question. I didn't hear any suggestions like that on the EP. It could be changed, but I wonder if we could have some restrictions or conventions for using it. Any suggestion?

JoelSpeed replied:
I think namespacing under machine.openshift.io would probably make sense, but we may already have labels that are similar

One convention is not to use underscores, it would be more idiomatic to use a hyphen.

Remind me what the zone type and groups are again? Are they anything consumed by Machine API or is this just grouping of MachineSets?

mtulio replied:
> Remind me what the zone type and groups are again? Are they anything consumed by Machine API or is this just grouping of MachineSets?

They are not consuming internally (for any OCP component). It was created to group the nodes (created by MachineSets) helping customers to create their applications and selecting workloads for each location in LZ.

Both are attributes from the [`AvailabilityZone` structure](https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_AvailabilityZone.html):

- `zone_type`: Availability Zone type, in this case, will always be `local-zone`
- `zone_group`: Each zone has its own group. The border group of Local Zone is the metropolitan region to which the zone belongs. For example in Los Angeles, there are two local zones, `us-west-2-lax-1a` and `us-west-2-lax-1b`, both are part of the border group `us-west-2-lax-1`

```
$ aws --region us-west-2 ec2 describe-availability-zones | jq -r '.AvailabilityZones[] | select(.GroupName == "us-west-2-lax-1")'
{
  "State": "available",
  "OptInStatus": "opted-in",
  "Messages": [],
  "RegionName": "us-west-2",
  "ZoneName": "us-west-2-lax-1a",
  "ZoneId": "usw2-lax1-az1",
  "GroupName": "us-west-2-lax-1",
  "NetworkBorderGroup": "us-west-2-lax-1",
  "ZoneType": "local-zone",
  "ParentZoneName": "us-west-2b",
  "ParentZoneId": "usw2-az2"
}
{
  "State": "available",
  "OptInStatus": "opted-in",
  "Messages": [],
  "RegionName": "us-west-2",
  "ZoneName": "us-west-2-lax-1b",
  "ZoneId": "usw2-lax1-az2",
  "GroupName": "us-west-2-lax-1",
  "NetworkBorderGroup": "us-west-2-lax-1",
  "ZoneType": "local-zone",
  "ParentZoneName": "us-west-2d",
  "ParentZoneId": "usw2-az4"
}
```

mtulio replied:
@JoelSpeed [I fixed](patch) the labels added to the machine.openshift.io namespace. IMHO it's a great idea, thanks for pointing it out. Let me know if it fits what you were thinking.

[patch]:https://github.com/openshift/installer/compare/72f58ca741fe1262dbf79c4acef6b138a67af89a..ee094569c1e9b59a55d694f6d83811e57c1c8b87

mtulio commented on pkg/asset/machines/aws/machines.go:
```diff
@@ -153,61 +168,74 @@ func Machines(clusterID string, region string, subnets map[string]string, pool *
 
-	if subnet == "" {
+	subnetName := fmt.Sprintf("%s-private-%s", in.clusterID, in.zone)
+	publicSubnet := !in.privateSubnet
+	if publicSubnet {
+		config.PublicIP = &publicSubnet
```
I will adjust when the comment above about pointer deprecation has been resolved. Thanks for raising it.

mtulio replied:
Fixed on [this review](https://github.com/openshift/installer/compare/cff35ba7dfe2b78984f49b73bfc20d11419fbe3c..c4d652e3d83bb4821eff4ed9b0a834c69c22ff1e)

patrickdillon commented:
>     * [ ]  Installer @patrickdillon

@sadasu will also be helping review this for the installer.


>     * [ ]  Network Team: Cluster network MTU Changes (@ ?)
> 
> @patrickdillon could you please advise anyone from Network team?

@abhat I think for cluster networking

mtulio commented:
Just successfull created a cluster after the recent changes (rebase + PR review):

```bash
$ OPENSHIFT_INSTALL_RELEASE_IMAGE_OVERRIDE="quay.io/openshift-release-dev/ocp-release:4.13.0-ec.3-x86_64" \
>   $INSTALLER create cluster --dir $INSTALL_DIR --log-level=debug
(...)
DEBUG Time elapsed per stage:                      
DEBUG            cluster: 4m0s                     
DEBUG          bootstrap: 38s                      
DEBUG Bootstrap Complete: 13m3s                    
DEBUG                API: 1m35s                    
DEBUG  Bootstrap Destroy: 50s                      
DEBUG  Cluster Operators: 11m35s                   
INFO Time elapsed: 30m41s 

$ oc get clusterversion
NAME      VERSION       AVAILABLE   PROGRESSING   SINCE   STATUS
version   4.13.0-ec.3   True        False         4h17m   Cluster version is 4.13.0-ec.3

$ oc get machines -n openshift-machine-api
NAME                                        PHASE     TYPE          REGION      ZONE               AGE
ocp-lz5-rxbtv-edge-us-east-1-nyc-1a-bnfgj   Running   c5d.2xlarge   us-east-1   us-east-1-nyc-1a   41m
ocp-lz5-rxbtv-master-0                      Running   m6i.xlarge    us-east-1   us-east-1a         45m
ocp-lz5-rxbtv-master-1                      Running   m6i.xlarge    us-east-1   us-east-1b         45m
ocp-lz5-rxbtv-master-2                      Running   m6i.xlarge    us-east-1   us-east-1c         45m
ocp-lz5-rxbtv-worker-us-east-1a-pg9w6       Running   m6i.xlarge    us-east-1   us-east-1a         41m
ocp-lz5-rxbtv-worker-us-east-1b-htfkg       Running   m6i.xlarge    us-east-1   us-east-1b         41m

$ oc get machinesets -n openshift-machine-api
NAME                                  DESIRED   CURRENT   READY   AVAILABLE   AGE
ocp-lz5-rxbtv-edge-us-east-1-nyc-1a   1         1         1       1           45m
ocp-lz5-rxbtv-worker-us-east-1a       1         1         1       1           45m
ocp-lz5-rxbtv-worker-us-east-1b       1         1         1       1           45m
ocp-lz5-rxbtv-worker-us-east-1c       1         1         1       1           45m

$ oc get nodes
NAME                           STATUS   ROLES                  AGE   VERSION
ip-10-0-143-235.ec2.internal   Ready    edge,worker            34m   v1.26.0+9eb81c2
ip-10-0-57-48.ec2.internal     Ready    control-plane,master   44m   v1.26.0+9eb81c2
ip-10-0-60-110.ec2.internal    Ready    worker                 38m   v1.26.0+9eb81c2
ip-10-0-66-79.ec2.internal     Ready    worker                 38m   v1.26.0+9eb81c2
ip-10-0-75-195.ec2.internal    Ready    control-plane,master   45m   v1.26.0+9eb81c2
ip-10-0-80-70.ec2.internal     Ready    control-plane,master   44m   v1.26.0+9eb81c2
ip-10-0-95-146.ec2.internal    Ready    worker                 35m   v1.26.0+9eb81c2

$ oc get co -w
NAME                                       VERSION       AVAILABLE   PROGRESSING   DEGRADED   SINCE   MESSAGE
authentication                             4.13.0-ec.3   True        False         False      26m     
baremetal                                  4.13.0-ec.3   True        False         False      41m     
cloud-controller-manager                   4.13.0-ec.3   True        False         False      44m     
cloud-credential                           4.13.0-ec.3   True        False         False      45m     
cluster-autoscaler                         4.13.0-ec.3   True        False         False      40m     
config-operator                            4.13.0-ec.3   True        False         False      42m     
console                                    4.13.0-ec.3   True        False         False      29m     
control-plane-machine-set                  4.13.0-ec.3   True        False         False      41m     
csi-snapshot-controller                    4.13.0-ec.3   True        False         False      41m     
dns                                        4.13.0-ec.3   True        False         False      41m     
etcd                                       4.13.0-ec.3   True        False         False      40m     
image-registry                             4.13.0-ec.3   True        False         False      35m     
ingress                                    4.13.0-ec.3   True        False         False      36m     
insights                                   4.13.0-ec.3   True        False         False      35m     
kube-apiserver                             4.13.0-ec.3   True        False         False      29m     
kube-controller-manager                    4.13.0-ec.3   True        False         False      39m     
kube-scheduler                             4.13.0-ec.3   True        False         False      38m     
kube-storage-version-migrator              4.13.0-ec.3   True        False         False      42m     
machine-api                                4.13.0-ec.3   True        False         False      38m     
machine-approver                           4.13.0-ec.3   True        False         False      41m     
machine-config                             4.13.0-ec.3   True        False         False      41m     
marketplace                                4.13.0-ec.3   True        False         False      41m     
monitoring                                 4.13.0-ec.3   True        False         False      30m     
network                                    4.13.0-ec.3   True        False         False      44m     
node-tuning                                4.13.0-ec.3   True        False         False      41m     
openshift-apiserver                        4.13.0-ec.3   True        False         False      29m     
openshift-controller-manager               4.13.0-ec.3   True        False         False      37m     
openshift-samples                          4.13.0-ec.3   True        False         False      35m     
operator-lifecycle-manager                 4.13.0-ec.3   True        False         False      41m     
operator-lifecycle-manager-catalog         4.13.0-ec.3   True        False         False      42m     
operator-lifecycle-manager-packageserver   4.13.0-ec.3   True        False         False      36m     
service-ca                                 4.13.0-ec.3   True        False         False      42m     
storage                                    4.13.0-ec.3   True        False         False      41m     

$ oc get network cluster -o json | jq .status
{
  "clusterNetwork": [
    {
      "cidr": "10.132.0.0/14",
      "hostPrefix": 23
    }
  ],
  "clusterNetworkMTU": 1200,
  "networkType": "OVNKubernetes",
  "serviceNetwork": [
    "172.30.0.0/16"
  ]
}


```

mtulio commented:
I am still trying to understand the relationship with those failures with this change, specially the job `ci/prow/e2e-aws-ovn` which is always failing in this PR, the API never goes up.

It should not impact in any OVN job as the file will never be created when there is no Local Zone subnet on the install-config. Example of installing in existing VPC with regular zones only (3 zones):

```bash
 $ echo ${#SUBNETS[*]}
6

$ export INSTALLER=./openshift-install
$ export RELEASE="quay.io/openshift-release-dev/ocp-release:4.13.0-ec.3-x86_64"
$ 
$ cp $INSTALL_DIR/install-config.yaml $INSTALL_DIR/install-config.yaml-bkp
$ OPENSHIFT_INSTALL_RELEASE_IMAGE_OVERRIDE="$RELEASE" \
>   $INSTALLER create manifests --dir $INSTALL_DIR
INFO Consuming Install Config from target directory 
INFO Manifests created in: ocp-lz5-2/manifests and ocp-lz5-2/openshift 
$ 
$ # Review if MTU patch has been created
$ ls -ls $INSTALL_DIR/manifests/cluster-network-*
12 -rw-r-----. 1 mtulio mtulio 9607 Feb 15 23:45 ocp-lz5-2/manifests/cluster-network-01-crd.yml
 4 -rw-r-----. 1 mtulio mtulio  273 Feb 15 23:45 ocp-lz5-2/manifests/cluster-network-02-config.yml
$ cat $INSTALL_DIR/manifests/cluster-network-03-config.yml
cat: ocp-lz5-2/manifests/cluster-network-03-config.yml: No such file or directory

$ cat $INSTALL_DIR/manifests/cluster-network-02-config.yml
apiVersion: config.openshift.io/v1
kind: Network
metadata:
  creationTimestamp: null
  name: cluster
spec:
  clusterNetwork:
  - cidr: 10.132.0.0/14
    hostPrefix: 23
  externalIP:
    policy: {}
  networkType: OVNKubernetes
  serviceNetwork:
  - 172.30.0.0/16
status: {}


$ grep subnets -A 7 $INSTALL_DIR/install-config.yaml-bkp
    subnets:
    - subnet-0412a2a25f7c41deb
    - subnet-0a04a5765a3f99b3f
    - subnet-0cc38e214eb1ed884
    - subnet-0ff38a3ff47608aeb
    - subnet-0e71abe2b9bfcffc9
    - subnet-0d8730487a32204af
pullSecret:
```

Also, the `e2e-aws-ovn-upi` passed the last test and failed now again.

/test e2e-aws-ovn
/test e2e-aws-ovn-upi

abhat commented:
@jcaamano @trozet review for mtu impact.

sadasu commented on pkg/asset/installconfig/aws/metadata.go:
```diff
@@ -74,6 +75,21 @@ func (m *Metadata) AvailabilityZones(ctx context.Context) ([]string, error) {
+// subnets that the cloud-provider logic considers to be edge
+// (i.e. Local Zone).
+func (m *Metadata) EdgeSubnets(ctx context.Context) (map[string]Subnet, error) {
+	m.mutex.Lock()
```
Regarding the mutex locking, I see this method is doing what all other methods are doing, but wouldn't it be better to grab the mutex lock inside populateSubnets() and remove it from all the functions that are calling it.

mtulio replied:
That's a good point. IMHO sharing this mutex with the session is not a better approach in this case. The `?Subnets()` functions are called from different points with the goal of getting the list of subnets that should be populated at once by `populateSubnets()`. The mutex `m.mutex` is being used also by `Session()` and called from many points from the code. Both should have different goals, although `?Subnets()` (`populateSubnets()`) also consumes the session.

I am introducing a mutex for `populateSubnets()`, used by  `?Subnets()`s, and setting a flag `subnetsPopulated` to avoid calling the `subnets()` many times. It also will help to prevent, eventually, getting into the case of returning empty `EdgeSubnets` when the `(Public||Private)Subnets` has been populated. I think the intention here is to check if the discovery was already been called.

Let me know your thoughts, @sadasu and @patrickdillon .

Reference [patch](patch).

[patch]:https://github.com/openshift/installer/compare/ee094569c1e9b59a55d694f6d83811e57c1c8b87..95bc39b33543d489b20c2da0cf3e38ef6f10dc3f#diff-60605f1bbbaa381ef60e730158a166623607b32eae21f481509e098bd24ff5ba

sadasu commented on pkg/asset/installconfig/aws/subnet.go:
```diff
@@ -23,13 +26,47 @@ type Subnet struct {
+	// The valid values are availability-zone and local-zone.
+	ZoneType string
+
+	// ZoneGroup is the group of subnet's availability zone is part of.
```
Reword to make this clear.

mtulio replied:
[Done](patch)

[patch]:https://github.com/openshift/installer/compare/ee094569c1e9b59a55d694f6d83811e57c1c8b87..95bc39b33543d489b20c2da0cf3e38ef6f10dc3f

sadasu commented on pkg/asset/installconfig/aws/subnet.go:
```diff
@@ -99,38 +139,59 @@ func subnets(ctx context.Context, session *session.Session, region string, ids [
+		if meta.ZoneType == localZoneType {
+			// Local Zones is supported only in Public subnets
+			if !meta.Public {
+				return subnetGroups, errors.Errorf("Local Zones subnets must be associated with public route tables: subnet %s from availability zone %s[%s] is public[%v]", id, meta.Zone, meta.ZoneType, meta.Public)
```
Do we want to return here or skip adding this subnet's metadata to subnetGroups.Edge and continue with  processing the next subnet id?

mtulio replied:
Yes, return here.
> Local Zones is supported only in Public subnets

This is what we are proposing and testing for Phase-1.

mtulio commented:
PR was rebased after a conflict with https://github.com/openshift/installer/pull/6814 .
@patrickdillon @r4f4 on #6814 we are doubling the size when SNO, using a function that I am removing in favor of using the AWS defaults package with more flexibility to discover instances for Local Zones. There is [a comment](comment-6814) referencing AWS does not support sizes 8 CPU with 16GiB of RAM, the compute-optimized supports it, you can check [here some options](instances-8x16). FYI: I am keeping the same logic introduced on 6814 (8x32), but moving to the platform/aws defaults package.

Please let me know your thoughts.

[comment-6814]:https://github.com/openshift/installer/pull/6371/files#diff-7abecca62150eb15cb02e930cd045f4d6782f3f2b3a933d3cb67c89bc6369026R52-R57
[instances-8x16]:https://instances.vantage.sh/?min_memory=15&min_vcpus=8&filter=c6a.2xlarge|c5a.2xlarge|c6i.2xlarge|c5.2xlarge|c5ad.2xlarge|c5d.2xlarge|c6id.2xlarge|c6in.2xlarge&compare_on=true&selected=c6a.2xlarge,c5a.2xlarge,c6i.2xlarge,c5.2xlarge,c5ad.2xlarge,c5d.2xlarge,c6id.2xlarge,c6in.2xlarge

mtulio commented:
@patrickdillon @sadasu comments addressed. Would you mind taking a look? We are targeting this PR to the current sprint, FF 4.13. If you have any questions, feel free to call me on Slack. I am fully focused on addressing your reviews/questions.

r4f4 commented:
> Also, the `e2e-aws-ovn-upi` passed the last test and failed now again.

aws-ovn-upi was fixed in https://github.com/openshift/installer/pull/6887

mtulio commented on pkg/asset/installconfig/aws/availabilityzones.go:
```diff
@@ -9,8 +9,14 @@ import (
+	// ZoneType can be 'availability-zone', 'local-zone' or 'wavelength-zone'.
+	availabilityZoneType = "availability-zone"
+	localZoneType        = "local-zone"
+)
```
@r4f4 great point. [Moved](https://github.com/openshift/installer/compare/21ddbd5f91fb644069d0cf90fb05d3eee15b8bcf..03221d8932cf343ee4e9ea32b3e5342a4bb710ae) to types package. Also renamed the "Regular" to "Availability", following the AWS standard. Thanks.

mtulio commented:
Triaging of job failures, I think we can group it (based on the last execution):

1) Failure: building libvirt-installer
~~~
  * could not run steps: step libvirt-installer failed: could not get build libvirt-installer: builds.build.openshift.io "libvirt-installer" not found 
~~~
Jobs impacted:
- [ci/prow/e2e-aws-ovn](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_installer/6371/pull-ci-openshift-installer-master-e2e-aws-ovn/1630220573508374528)
- [ci/prow/e2e-aws-ovn-imdsv2](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_installer/6371/pull-ci-openshift-installer-master-e2e-aws-ovn-imdsv2/1630220573609037824)
- [ci/prow/e2e-aws-ovn-upi](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_installer/6371/pull-ci-openshift-installer-master-e2e-aws-ovn-upi/1630220574229794816)

2) Failure: bootstrap timeout

Jobs impacted:
- [ci/prow/e2e-aws-ovn-fips](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_installer/6371/pull-ci-openshift-installer-master-e2e-aws-ovn-fips/1630220573567094784)
- [ci/prow/okd-scos-e2e-aws-ovn](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_installer/6371/pull-ci-openshift-installer-master-okd-scos-e2e-aws-ovn/1630220581808902144)
- [ci/prow/e2e-aws-ovn-upgrade](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_installer/6371/pull-ci-openshift-installer-master-e2e-aws-ovn-upgrade/1630220573781004288) : `error: timed out waiting for the condition on clusteroperators/etcd`
- [ci/prow/e2e-aws-ovn-workers-rhel8](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_installer/6371/pull-ci-openshift-installer-master-e2e-aws-ovn-workers-rhel8/1630220574796025856): `msg=EtcdMembersDegraded: 2 of 3 members are available, etcd-bootstrap is unhealthy`
- [ci/prow/okd-e2e-aws-ovn](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_installer/6371/pull-ci-openshift-installer-master-okd-e2e-aws-ovn/1630220579330068480): `msg=EtcdMembersDegraded: 2 of 3 members are available, etcd-bootstrap is unhealthy`

3) Failed: Flaky?

Jobs impacted:
- [ci/prow/e2e-aws-ovn-proxy](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_installer/6371/pull-ci-openshift-installer-master-e2e-aws-ovn-proxy/1630220573659369472)
- [ci/prow/e2e-aws-ovn-single-node](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_installer/6371/pull-ci-openshift-installer-master-e2e-aws-ovn-single-node/1630220573734866944)

/test e2e-aws-ovn-proxy
/test e2e-aws-ovn-single-node

4) Failures: image registry issues. Sample

~~~
ERRO[2023-02-27T20:15:25Z] Some steps failed:                           
ERRO[2023-02-27T20:15:25Z] 
  * could not run steps: step [release-inputs:latest] failed: could not create snapshot imagestream origin/scos-4.13 for release latest: Timeout: request did not complete within requested timeout - context deadline exceeded
  * could not run steps: step [release-inputs:initial] failed: could not create snapshot imagestream origin/scos-4.13 for release initial: Timeout: request did not complete within requested timeout - context deadline exceeded 
~~~

Jobs impacted:
- [ci/prow/images](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_installer/6371/pull-ci-openshift-installer-master-images/1630267592885145600)
- [ci/prow/okd-images](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_installer/6371/pull-ci-openshift-installer-master-okd-images/1630267592927088640)
- [ci/prow/okd-scos-images](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_installer/6371/pull-ci-openshift-installer-master-okd-scos-images/1630267592956448768)

I didn't find connections between the current changes. Open to hearing from more experienced folks on these CI jobs.

I also get informed that we are having CI issues (performance problems on the build farm control planes) since last Thursday. :(

/retest-required

mtulio commented:
/unhold

mtulio commented:
I've noticed that the CI instability should be fixed on the EOD/Monday.
/retest

patrickdillon reviewed (commented):
/approve

In general, this LGTM. I need to spend a bit more time reviewing the last commit, but I suspect we should be able to tag this soon.

mtulio commented:
/test e2e-aws-ovn

mtulio commented on docs/user/aws/install_existing_vpc_local-zones.md:
```diff
@@ -0,0 +1,422 @@
+    - `zone_group=<Local Zone Group>`
+
+Finally, the Machine Sets created by the `edge` compute pool have `NoSchedule` taint to avoid the
+regular workloads spread out on those machines, and only user workloads will be allowed to run
+when the tolerations are defined on the pod spec.
```
For sure, we created it on the HC blog post[1]. I will paste the full example here, and adjust when updating the doc:
```
cat << EOF | oc create -f -
apiVersion: v1
kind: Namespace
metadata:
  name: lz-apps
---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: lz-app-nyc-1
  namespace: lz-apps
spec:
  selector:
    matchLabels:
      app: lz-app-nyc-1
  replicas: 1
  template:
    metadata:
      labels:
        app: lz-app-nyc-1
        zone_group: ${ZONE_GROUP_NAME}
    spec:
      nodeSelector:
        zone_group: ${ZONE_GROUP_NAME}
      tolerations:
      - key: "node-role.kubernetes.io/edge"
        operator: "Equal"
        value: ""
        effect: "NoSchedule"
      containers:
        - image: openshift/origin-node
          command:
           - "/bin/socat"
          args:
            - TCP4-LISTEN:8080,reuseaddr,fork
            - EXEC:'/bin/bash -c \"printf \\\"HTTP/1.0 200 OK\r\n\r\n\\\"; sed -e \\\"/^\r/q\\\"\"'
          imagePullPolicy: Always
          name: echoserver
          ports:
            - containerPort: 8080
---
apiVersion: v1
kind: Service 
metadata:
  name:  lz-app-nyc-1 
  namespace: lz-apps
spec:
  ports:
    - port: 80
      targetPort: 8080
      protocol: TCP
  type: NodePort
  selector:
    app: lz-app-nyc-1
EOF
```

[1] https://docs.google.com/document/d/13C-FPCgYEMsr2cyVV_pG5mpl03ZjfvXrV8d2uSJVms0/edit

mtulio replied:
Added a new section to the `Use Case` section. Thanks for the feedback.

mtulio commented on docs/user/aws/install_existing_vpc_local-zones.md:
```diff
@@ -0,0 +1,422 @@
+
+To install a cluster in existing VPC with Local Zone subnets, you should provision the network resources and then add the subnet IDs to the `install-config.yaml`.
+
+## Pre-requisites <a name="pre-requisites"></a>
```
yeah, I like to define the index due we don't need to keep long references, or can change the section name without needing to change the links, avoid broking it.
Thanks for your review, I will adjust it and tag you again.

mtulio commented:
> /approve
> 
> In general, this LGTM. I need to spend a bit more time reviewing the last commit, but I suspect we should be able to tag this soon.

@patrickdillon thanks for the review. Documentation fixed and PR rebased with the latest conflict. Creating a cluster now with the most recent changes will update here soon.

I am awaiting your final review! Thanks again.

mtulio commented:
Having issues with deployment after the rebase, the instanceType is not correctly added to the MachineSet manifest, failing to create the machine. I will fix it asap.
- Manifest generated by installer:
```
          instanceType: ""
```
- machine event:
```
$ oc  describe machine ocp-lz67-c6pvj-edge-us-east-1-scl-1a-bgt7h -n openshift-machine-api
Events:
  Type     Reason        Age   From           Message
  ----     ------        ----  ----           -------
  Warning  FailedCreate  37m   awscontroller  ocp-lz67-c6pvj-edge-us-east-1-scl-1a-bgt7h: reconciler failed to Create machine: failed to launch instance: error launching instance: Invalid value '' for InstanceType.
```

/hold

sadasu commented:
@mtulio thanks for addressing my comments. Will take a look again after you remove the hold.

mtulio commented:
> Having issues with deployment after the rebase, the instanceType is not correctly added to the MachineSet manifest, failing to create the machine. I will fix it asap.
> 
> /hold

I fell into one interesting situation here: tl;dr I added more instance types to the preferred pool to cover the Local Zones locations recently added by AWS.

--

In my labs I am choosing randomly[A] (as expected to test different locations) the Local Zone to create the subnet, my last cluster got the `us-east-1-scl-1a` (Santiago, Chile-> Location recently added). That zone does not have any of the preferred instances from the intersection mapped at the begging of the EP/research. It seems AWS does not keep one strict pattern (or it does have internal reasons) on new locations. What I can see from the new locations was that AWS is supporting instances not available in the US, except us-west-2, starting from 2xlarge for the families we support (not considering t3). I exported the matrix of availability across all zones to that gist: https://gist.github.com/mtulio/c98aa15128a7becb06a372f00d824c42#file-aws-local-zones-ec2-offerings-txt

Having said that, when I created the cluster I didn't take attention to the "WARNING" message I am logging when choosing the instanceType for a given location (`us-east-1-scl-1a`), keeping the default not supported, then the MachineSet will not be able to create the instance, nothing that affects the installer. It also can be fixed on Day-2 by patching the MachineSet. This warning is also the default behavior for "worker compute pool".

~~~
INFO Consuming Install Config from target directory 
WARNING unable to select instanceType on the zone[us-east-1-scl-1a] from the preferred list: [m6i.xlarge m5.xlarge c5d.2xlarge]. You must update the MachineSet manifest: no instance type found for the zone constraint 
INFO Manifests created in: ocp-lz67-19/manifests and ocp-lz67-19/openshift 
~~~

So based on that information, I am keeping those decisions:

1) if AWS add new Locations AND there are no instance type available on that zone from the preferred list of the installer (static on code), we will not fail, but keep the warning (above). So the user can define it on the edge compute pool[B]. This is the current behavior of the worker compute pool.
-> Why not fail? Because we can fall into a lock situation when the user would like to use XPTO instance type that is available in Local Zone A but not in B (as you can see on the matrix, the AWS does not make our lives easy). He can read the warning, ignore it, and update the MachineSet manifest on the create manifest stage.
2) Increase the preferred list to the families supported by us, increasing the size. So we can increase the coverage, proactively decrease manual steps, and organically converge to the same used by default on workers when AWS will update the HW of those zones. Example of the current list [m6i.xlarge, m5.xlarge] and new [m6i.xlarge, m5.xlarge, r5.xlarge, c5.2xlarge, m5.2xlarge, c5d.2xlarge] preferred list.
-> Why that order? Newest generation and price

[A] cmd used to select randomly the zones (cc @vandercool)

~~~bash
AZ_NAME=$(aws --region $CLUSTER_REGION ec2 describe-availability-zones \
  --filters Name=opt-in-status,Values=opted-in Name=zone-type,Values=local-zone \
  | jq -r .AvailabilityZones[].ZoneName | shuf | head -n1)
~~~
[B] custom instanceType on install-config.yaml

~~~yaml
compute:
- name: edge
  platform:
    aws:
      type: m5.2xlarge
~~~

/unhold

mtulio commented:
@patrickdillon I found this[1] nit (`m4-large`) while working here. I tried to add it here but would be better to a new PR.

[changes](https://github.com/openshift/installer/compare/d1529caaac932fd0c6e3e8838541503107ecbd55..469b55370727894085893e80eb2de0819f1877cd)
~~~
$ openshift-install-linux-4.12.0 explain  installconfig.controlPlane.platform.aws.type
KIND:     InstallConfig
VERSION:  v1

RESOURCE: <string>
  InstanceType defines the ec2 instance type. eg. m4-large

~~~

rvanderp3 commented:
/lgtm

rvanderp3 commented:
/hold

rvanderp3 commented:
/hold cancel

removing hold.  there appears to be an unrelated issue with reconciliation of control planes via the control plane machinset operator.  will open a bug and comment on this PR when it is open

mtulio commented:
The following Bugs have been created with today's investigation of failing `e2e-aws-ovn*` jobs. [1] is directly related, [2] not impacting for failures, but is changing the default AWS deployment.

[1] https://issues.redhat.com/browse/OCPBUGS-8106
[2] https://issues.redhat.com/browse/OCPBUGS-8109

mtulio commented on pkg/asset/machines/aws/machines.go:
```diff
@@ -37,19 +52,20 @@ func Machines(clusterID string, region string, subnets map[string]string, pool *
+			root:           &mpool.EC2RootVolume,
+			imds:           mpool.EC2Metadata,
+			userTags:       userTags,
+			privateSubnet:  false,
```
I think this flag should be true

patrickdillon commented on pkg/asset/machines/aws/machines.go:
```diff
@@ -37,19 +52,20 @@ func Machines(clusterID string, region string, subnets map[string]string, pool *
+			root:           &mpool.EC2RootVolume,
+			imds:           mpool.EC2Metadata,
+			userTags:       userTags,
+			publicSubnet:   false,
```
Why add `publicSubnet` as an arg to the function if it is just a static value of false? Why not just use the `false` value within the function itself?

mtulio replied:
the `provider()`[1] function is used by `MachineSet`[2] which is not static.
[1] https://github.com/openshift/installer/pull/6371/commits/2e18401e10feb76d4ec25b6088253a4952049614#diff-e46d61c55e5e276e3c264d18cba0346777fe3e662d0180a173001b8282af7c6eR172
[2] https://github.com/openshift/installer/blob/2e18401e10feb76d4ec25b6088253a4952049614/pkg/asset/machines/aws/machinesets.go

https://github.com/openshift/installer/pull/6371/commits/2e18401e10feb76d4ec25b6088253a4952049614

mtulio replied:
`MachineSets()` are responsible to generate the worker (and not edge) MachineSets manifests

mtulio commented:
Passed from the last installer failures with the latest changes (publicIP flag). Now e2e is flaking with one test. There are jobs flaking too with terraform. Retesting it.



Terraform error creating resources:

- [ci/prow/e2e-aws-ovn-upgrade](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_installer/6371/pull-ci-openshift-installer-master-e2e-aws-ovn-upgrade/1631301289574928384)
- [ci/prow/e2e-aws-ovn-workers-rhel8](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_installer/6371/pull-ci-openshift-installer-master-e2e-aws-ovn-workers-rhel8/1631301290183102464)

e2e error:

- [ci/prow/e2e-aws-ovn](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_installer/6371/pull-ci-openshift-installer-master-e2e-aws-ovn/1631301289390379008)

~~~
Suite run returned error: 2 fail, 1352 pass, 1628 skip (42m7s)
error: 2 fail, 1352 pass, 1628 skip (42m7s)
      "Name": "[sig-storage] Mounted volume expand [Feature:StorageProvider] Should verify mounted devices can be resized [Skipped:NoOptionalCapabilities] [Suite:openshift/conformance/parallel] [Suite:k8s]",
      "Risk": {
        "Level": {
          "Name": "Medium",
          "Level": 50
        },
      }, 
~~~

- [ci/prow/e2e-aws-ovn-single-node](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_installer/6371/pull-ci-openshift-installer-master-e2e-aws-ovn/1631301289390379008) 

~~~

Suite run returned error: 24 fail, 1296 pass, 1662 skip (1h10m47s)
error: 24 fail, 1296 pass, 1662 skip (1h10m47s)
~~~

- [ci/prow/e2e-aws-ovn-fips](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_installer/6371/pull-ci-openshift-installer-master-e2e-aws-ovn-fips/1631301289419739136)


~~~
      "Name": "[sig-storage] Mounted volume expand [Feature:StorageProvider] Should verify mounted devices can be resized [Skipped:NoOptionalCapabilities] [Suite:openshift/conformance/parallel] [Suite:k8s]",
      "Risk": {
        "Level": {
          "Name": "Low",
          "Level": 1
        },
~~~

/retest-required
/test e2e-aws-ovn-workers-rhel8
/test e2e-aws-ovn-upgrade
/test e2e-aws-ovn-single-node
/test e2e-aws-ovn-fips

rvanderp3 commented:
/lgtm

rvanderp3 commented:
> Passed from the last installer failures with the latest changes (publicIP flag). Now e2e is flaking with one test. There are jobs flaking too with terraform. Retesting it.
> 
> Terraform error creating resources:
> 
> * [ci/prow/e2e-aws-ovn-upgrade](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_installer/6371/pull-ci-openshift-installer-master-e2e-aws-ovn-upgrade/1631301289574928384)
> * [ci/prow/e2e-aws-ovn-workers-rhel8](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_installer/6371/pull-ci-openshift-installer-master-e2e-aws-ovn-workers-rhel8/1631301290183102464)
> 
> e2e error:
> 
> * [ci/prow/e2e-aws-ovn](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_installer/6371/pull-ci-openshift-installer-master-e2e-aws-ovn/1631301289390379008)
> 
> ```
> Suite run returned error: 2 fail, 1352 pass, 1628 skip (42m7s)
> error: 2 fail, 1352 pass, 1628 skip (42m7s)
>       "Name": "[sig-storage] Mounted volume expand [Feature:StorageProvider] Should verify mounted devices can be resized [Skipped:NoOptionalCapabilities] [Suite:openshift/conformance/parallel] [Suite:k8s]",
>       "Risk": {
>         "Level": {
>           "Name": "Medium",
>           "Level": 50
>         },
>       }, 
> ```
> 
> * [ci/prow/e2e-aws-ovn-single-node](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_installer/6371/pull-ci-openshift-installer-master-e2e-aws-ovn/1631301289390379008)
> 
> ```
> 
> Suite run returned error: 24 fail, 1296 pass, 1662 skip (1h10m47s)
> error: 24 fail, 1296 pass, 1662 skip (1h10m47s)
> ```
> 
> * [ci/prow/e2e-aws-ovn-fips](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_installer/6371/pull-ci-openshift-installer-master-e2e-aws-ovn-fips/1631301289419739136)
> 
> ```
>       "Name": "[sig-storage] Mounted volume expand [Feature:StorageProvider] Should verify mounted devices can be resized [Skipped:NoOptionalCapabilities] [Suite:openshift/conformance/parallel] [Suite:k8s]",
>       "Risk": {
>         "Level": {
>           "Name": "Low",
>           "Level": 1
>         },
> ```
> 
> /retest-required /test e2e-aws-ovn-workers-rhel8 /test e2e-aws-ovn-upgrade /test e2e-aws-ovn-single-node /test e2e-aws-ovn-fips

these failures appear to be unrelated to the changes in this PR

patrickdillon commented:
/skip

rvanderp3 commented:
@patrickdillon I think we are stuck on https://issues.redhat.com/browse/OCPBUGS-8092

patrickdillon commented:
/skip

patrickdillon commented:
> @patrickdillon I think we are stuck on https://issues.redhat.com/browse/OCPBUGS-8092

Several other e2e-aws jobs, like imdsv2, shared vpc, proxy have succeeded. Plenty of confirmation aws works.

/override ci/prow/e2e-aws-ovn

patrickdillon commented:
/skip

patrickdillon commented:
/refresh

patrickdillon commented:
/skip

patrickdillon commented:
/refresh

patrickdillon commented:
/skip

patrickdillon commented:
/override ci/prow/e2e-aws-ovn-fip
/override ci/prow/e2e-aws-ovn-single-node 
/override ci/prow/e2e-aws-ovn-workers-rhel8
/override ci/prow/okd-e2e-aws-ovn
/override ci/prow/okd-scos-e2e-aws-ovn
/override ci/prow/okd-scos-e2e-aws-upgrade

patrickdillon commented:
/refresh

patrickdillon commented:
/override ci/prow/e2e-aws-ovn-fips
/override ci/prow/e2e-aws-ovn-single-node
/override ci/prow/e2e-aws-ovn-workers-rhel8
/override ci/prow/okd-e2e-aws-ovn
/override ci/prow/okd-scos-e2e-aws-ovn
/override ci/prow/okd-scos-e2e-aws-upgrade

patrickdillon commented:
/override ci/prow/okd-e2e-aws-ovn-upgrade

sdodson commented:
/cherry-pick release-4.13
